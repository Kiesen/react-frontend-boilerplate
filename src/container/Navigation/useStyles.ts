import { makeStyles, Theme } from '@material-ui/core/styles';

/**
 * Mui styling for the navigation container component.
 *
 * @exports StylesHook
 */
export default makeStyles((theme: Theme) => ({
  toolbar: {
    minHeight: theme.minNavHeight,
    maxWidth: theme.maxWidth,
    padding: '0 10px',
    margin: 'auto',
    width: '100%',
  },
  title: {
    flexGrow: 1,
  },
  link: {
    textDecoration: 'none',
    color: '#fff',
  },
}));
