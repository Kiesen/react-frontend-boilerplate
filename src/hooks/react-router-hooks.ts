/**
 * All hooks are related to issue 6430 of the react router repository:
 * https://github.com/ReactTraining/react-router/issues/6430
 */
import { useContext } from 'react';
import { __RouterContext as RouterContext } from 'react-router';

/**
 * Custom hook that returns the react router object.
 * Uses the react context hook with the react router
 * context as an argument.
 *
 * @return {Object} router - router object
 */
export const useRouter = () => {
  return useContext(RouterContext);
};

/**
 * Custom hook that just returns the
 * react router params.
 *
 * @return {Object} params - param object
 */
export const useParams = () => {
  const { match } = useRouter();
  return match.params;
};

/**
 * Custom hook that returns the router location object
 * and a custom function that uses push or replace
 * to change the route.
 *
 * @return an object containing location and navigate.
 */
export const useLocation = () => {
  const { location, history } = useRouter();

  /**
   * Custom function that uses the react router history to
   * either push the new route on the history stack
   * or replace the current route.
   *
   * @param {string} to - the new location
   * @param {boolean} [replace] - should replace be used
   *
   */
  const navigate = (to: string, replace = false) => {
    if (replace) {
      history.replace(to);
    } else {
      history.push(to);
    }
  };

  return {
    location,
    navigate,
  };
};
