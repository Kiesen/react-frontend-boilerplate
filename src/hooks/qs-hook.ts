import qs from 'qs';

import { useRouter } from 'hooks/react-router-hooks';

interface Props {
  [key: string]: string | null;
}
type SetQueryString = () => (props: Props) => void;

/**
 * Custom hook that returns a callback.
 * The callback function takes a one dimensional object as an argument,
 * creates a new query string and pushes the query string to the
 * history stack.
 *
 * Important note: If you want to remove any keys from the query string
 * just provide the key with a null value e.g. { x: null }
 *
 * @return callback with following signature (queryObject) => void
 */
export const useSetQueryString: SetQueryString = () => {
  const { location, history } = useRouter();

  return (props: Props) => {
    const queryString = qs.stringify(
      {
        ...qs.parse(location.search.slice(1)),
        ...qs.parse(qs.stringify(props, { strictNullHandling: true }), {
          strictNullHandling: true,
        }),
      },
      { addQueryPrefix: true, skipNulls: true }
    );
    history.push(`${location.pathname}${queryString}${location.hash}`);
  };
};
