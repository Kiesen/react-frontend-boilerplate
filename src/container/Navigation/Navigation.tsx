import React, { FC, SyntheticEvent } from 'react';
import { useTranslation } from 'react-i18next';
import { Link } from 'react-router-dom';

import Typography from '@material-ui/core/Typography';
import Toolbar from '@material-ui/core/Toolbar';
import Button from '@material-ui/core/Button';
import AppBar from '@material-ui/core/AppBar';
import Grid from '@material-ui/core/Grid';

import useStyles from 'container/Navigation/useStyles';
import { useLocation } from 'hooks/react-router-hooks';
import { AUTH_CREDENTIALS } from 'constants/storageKeys';
import { LOGIN, BASE } from 'constants/routes';

/**
 * Simple navigation container component.
 * Handels the logout action and shows the app title.
 *
 * @returns a sticky navigation bar
 */
const Navigation: FC<{}> = () => {
  // CSS classes created by material ui
  const classes = useStyles();
  // i18n translations
  const { t } = useTranslation();
  // Get the history from the router context
  const { navigate } = useLocation();

  /**
   * This function handels the logout action.
   * First remove any auth related items, than
   * redirect to the login view.
   *
   * @param {SyntheticEvent} event - SyntheticEvent<HTMLButtonElement>
   */
  const logout = (event: SyntheticEvent<HTMLButtonElement>): void => {
    event.preventDefault();
    // -- TODO --
    // Depending on the auth service or the api you use the
    // auth logic must be updated.
    sessionStorage.removeItem(AUTH_CREDENTIALS);
    navigate(LOGIN);
  };

  return (
    <AppBar position="sticky">
      <Toolbar className={classes.toolbar}>
        <Grid container spacing={0}>
          <Grid item xs={6}>
            <Typography variant="h6" className={classes.title}>
              <Link className={classes.link} to={BASE}>
                react-frontend-boilerplate
              </Link>
            </Typography>
          </Grid>

          <Grid item xs={6}>
            <Grid container justify="flex-end">
              <Button variant="contained" color="secondary" onClick={logout}>
                {t('client:button.logout')}
              </Button>
            </Grid>
          </Grid>
        </Grid>
      </Toolbar>
    </AppBar>
  );
};

export default Navigation;
