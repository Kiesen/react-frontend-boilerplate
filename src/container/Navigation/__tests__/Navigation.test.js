import React from 'react';
import { shallow } from 'enzyme';

import * as hooks from 'hooks/react-router-hooks';
import Navigation from 'container/Navigation/Navigation';

describe('<Navigation />', () => {
  beforeEach(() => {
    jest.spyOn(hooks, 'useLocation').mockImplementation(() => ({
      navigate: jest.fn(),
    }));
  });

  it('renders correctly', () => {
    const wrapper = shallow(<Navigation />);
    expect(wrapper).toMatchSnapshot();
  });
});
