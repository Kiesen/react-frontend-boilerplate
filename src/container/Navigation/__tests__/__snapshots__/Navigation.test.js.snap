// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> renders correctly 1`] = `
<WithStyles(ForwardRef(AppBar))
  position="sticky"
>
  <WithStyles(ForwardRef(Toolbar))
    className="makeStyles-toolbar-1"
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      spacing={0}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={6}
      >
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-title-2"
          variant="h6"
        >
          <Link
            className="makeStyles-link-3"
            to="/"
          >
            react-frontend-boilerplate
          </Link>
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={6}
      >
        <WithStyles(ForwardRef(Grid))
          container={true}
          justify="flex-end"
        >
          <WithStyles(ForwardRef(Button))
            color="secondary"
            onClick={[Function]}
            variant="contained"
          >
            client:button.logout
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
